projects/storefrontlib/cms-components/product/product-list/container/product-list-component.service.ts

```diff
@@ -84,10 +85,26 @@ export class ProductListComponentService {
           state.params,
           state.queryParams
         );
-        this.search(criteria);
+        this.searchIfQueryHasChanged(criteria);
       })
     );
 
+  /**
+   * Search only if the previous search query does NOT match the new one.
+   * This prevents repeating product search calls for queries that already have loaded data.
+   */
+  protected searchIfQueryHasChanged(criteria: SearchCriteria) {
```

Comment:
[NITPICK] (no need to change, just sharing an idea for a more reactive-approach implementation, which unfortunately cannot be applied because it would be a formal breaking change)

To avoid of doing an imperative `subscribe()` and naming a method with `if` (`IfSomethingSomething`), I'd consider using more stream/reactive approach (e.g. using `withLatestFrom(this.currentSearchResultsQuery$)` and `filter(""newQuery !== oldQuery"")`:

```diff  
-      tap((state: ActivatedRouterStateSnapshot) => {
-        const criteria = this.getCriteriaFromRoute(
-          state.params,
-          state.queryParams
-        );
-        this.searchIfQueryHasChanged(criteria);
-      })
+     map((state: ActivatedRouterStateSnapshot) =>
+       this.getCriteriaFromRoute(
+          state.params,
+          state.queryParams
+        )
+      ),
+      withLatestFrom(this.currentSearchResultsQuery$),
+      filter(([newCriteria, currentQuery]) => newCriteria.query !== currentQuery)
+      tap(newCriteria => search(criteria))
```

and add:

```ts
protected currentSearchResultsQuery$ = productSearchService
  .getResults()
  .pipe(
    map(results => results?.currentQuery?.query?.value)
  );
```

---

All that said, it would change the return type of the `protected searchByRouting$` from `Observable<ActivatedRouterStateSnapshot>` to `Observable<[SearchCriteria, string]>`, which is a breaking change. Of course we're not using this observable anywhere to get it's value but only to trigger the side-effect of calling `search()`, but by the way we're changing the return type of the `protected` observable which formally is a breaking change.